### PR TITLE
networkd and sd-netlink: add support for Generic netlink And FooOverUDP  to IPIP tunnel.

### PR DESCRIPTION
**Pull request by ssahani**
This work add support to generic netlink to sd-netlink.
See https://lwn.net/Articles/208755/

networkd: add support FooOverUDP support to IPIP tunnel netdev
https://lwn.net/Articles/614348/

Example conf:

/lib/systemd/network/1-fou-tunnel.netdev
```
[NetDev]
Name=fou-tun
Kind=fou

[FooOverUDP]
Port=5555
Protocol=4

```

/lib/systemd/network/ipip-tunnel.netdev
```
[NetDev]
Name=ipip-tun
Kind=ipip

[Tunnel]
Independent=true
Local=10.65.208.212
Remote=10.65.208.211
FooOverUDP=true
FOUDestinationPort=5555
```

$ ip -d link show ipip-tun
```
5: ipip-tun@NONE: <POINTOPOINT,NOARP> mtu 1472 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ipip 10.65.208.212 peer 10.65.208.211 promiscuity 0
    ipip remote 10.65.208.211 local 10.65.208.212 ttl inherit pmtudisc encap fou encap-sport auto encap-dport 5555 noencap-csum noencap-csum6 noencap-remcsum numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535
```

### CODE REVIEW

**Comment by ssahani** [comment]
Ping.

File: configure.ac [hunk 1]
```diff
@@ -368,7 +368,10 @@ AC_CHECK_DECLS([IFLA_INET6_ADDR_GEN_MODE,
                 IFLA_BRPORT_LEARNING_SYNC,
                 IFLA_BR_VLAN_DEFAULT_PVID,
                 NDA_IFINDEX,
-                IFA_FLAGS],
+                IFA_FLAGS,
+                FOU_ATTR_REMCSUM_NOPARTIAL,
+                FOU_CMD_GET,
+                FOU_ENCAP_GUE],
```
Comment (keszybz) [inline configure.ac]: This is missing corresponding meson changes.

File: src/libsystemd/sd-netlink/generic-netlink-message.c [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        return 0;
+}
+
+int sd_generic_netlink_message_append_u8(sd_netlink_message *m, unsigned short type, uint8_t data) {
+        int r;
+
+        assert_return(m, -EINVAL);
+        assert_return(!m->sealed, -EPERM);
+
+        r = sd_netlink_message_append_data(m, type, &data, sizeof(uint8_t));
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/generic-netlink-message.c]: Those append functions can be just thin wrappers around sd_netlink_message_append_data:
```C
{
  return sd_netlink_message_append_data(m, type, &data, sizeof uint8_t);
}
```
— no need to duplicate all the checks.

File: src/libsystemd/sd-netlink/generic-netlink-message.c [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+int sd_generic_netlink_message_append_string(sd_netlink_message *m, unsigned short type, const char *data) {
+        size_t length;
+        int r;
+
+        assert_return(m, -EINVAL);
+        assert_return(!m->sealed, -EPERM);
+        assert_return(data, -EINVAL);
+
+        length = strlen(data);
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/generic-netlink-message.c]: Just do
```C
assert_return(data, -EINVAL);
return sd_netlink_message_append_data(m, type, data, strlen(data) + 1);
```
— no need to duplicate all other checks.

File: src/libsystemd/sd-netlink/generic-netlink-message.c [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        uint32_t len;
+
+        assert_return(rta, -EINVAL);
+        assert_return(tb, -EINVAL);
+
+        rt = RTA_DATA(rta);
+        len = RTA_PAYLOAD(rta);
+
+        for (; RTA_OK(rt, len); rt = RTA_NEXT(rt, len))
+                tb[rt->rta_type] = rt;
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/generic-netlink-message.c]: How do we know that `tb` has the right length? <strike>I think you should pass a `size_t` argument.</strike>
OK, so `tb` is always supposed to be (CTRL_ATTR_MAX + 1)-long. Then change the parameter type to `struct rtattr tb[CTRL_ATTR_MAX + 1]` so it's clear to the reader.

File: src/test/test-generic-netlink-manual.c [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+        assert_se(sd_rtnl_message_new_generic_netlink(type, FOU_CMD_ADD, FOU_GENL_VERSION, &m) >= 0);
+        assert_se(sd_generic_netlink_message_append_u16(m, FOU_ATTR_PORT, htobe16(port)) >= 0);
+        assert_se(sd_generic_netlink_message_append_u8(m, FOU_ATTR_TYPE, FOU_ENCAP_GUE) >= 0);
+        assert_se(sd_generic_netlink_message_append_u16(m, FOU_ATTR_AF, AF_INET) >= 0);
+        assert_se(sd_generic_netlink_message_append_u8(m, FOU_ATTR_IPPROTO, 4) >= 0);
+
+        r = sd_netlink_call(rtnl, m, 0, NULL);
+        if (r < 0)
+                printf("Failed to add FooOverUDP port: %d \t %s\n", port, strerror(-r));
```
Comment (keszybz) [inline src/test/test-generic-netlink-manual.c]: Just use log_error and %m.

File: src/test/test-generic-netlink-manual.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                }
+
+                printf("\n");
+        }
+        printf("\n");
+
+        assert_se((m = sd_netlink_message_unref(m)) == NULL);
+        assert_se((n = sd_netlink_message_unref(n)) == NULL);
+
+        return EXIT_SUCCESS;
```
Comment (keszybz) [inline src/test/test-generic-netlink-manual.c]: `return 0`. We use EXIT_* only for values which are (eventually) returned by `main()`.

File: man/systemd.netdev.xml [hunk 7]
```diff
@@ -747,8 +750,80 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>FooOverUDP=</varname></term>
+        <listitem>
+          <para>A boolean. Specifies whether <varname>FooOverUDP=</varname> tunnel is to be configured.
+           Default to false. For more detail information see
+          <ulink url="https://lwn.net/Articles/614348"> Foo over UDP</ulink></para>
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Extra space before "Foo".

File: man/systemd.netdev.xml [hunk 8]
```diff
@@ -747,8 +750,80 @@
...
+        </listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><varname>FooOverUDPEncapType=</varname></term>
+        <listitem>
+          <para>Please see FooOverUDP section's <varname>EncapType=</varname>.
```
Comment (keszybz) [inline man/systemd.netdev.xml]: This needs some explanation. How is this option different from EncapType=?

File: man/systemd.netdev.xml [hunk 9]
```diff
@@ -747,8 +750,80 @@
...
+          <para>The <varname>FooOverUDPDestinationPort=</varname> specifies the
+          UDP destination port for encapsulation. This field is mandotory and is not set
+          by default.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>FooOverUDPSourcePort=</varname></term>
+        <listitem>
+          <para>The <varname>FooOverUDPSourcePort=</varname> specifies the
+          UDP source port for encapsulation. Default to <varname>auto</varname> or <varname>0</varname> ,
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Default*s*. Extra space before comma at the end.

File: man/systemd.netdev.xml [hunk 10]
```diff
@@ -747,8 +750,80 @@
...
+    <para>The <literal>[FouOverUDP]</literal> section only applies for
+    netdevs of kind <literal>fou</literal> and accepts the
+    following keys:</para>
+
+    <variablelist class='network-directives'>
+      <varlistentry>
+        <term><varname>Protocol=</varname></term>
+        <listitem>
+          <para>The <varname>Protocol=</varname> specifies the protocol number of the
+          packets arriving at <varname>FooOverUDPPort=</varname>. This field is mandotory and is not set
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Typo in "mondotory".

File: man/systemd.netdev.xml [hunk 11]
```diff
@@ -747,8 +750,80 @@
...
+      </varlistentry>
+      <varlistentry>
+        <term><varname>EncapType=</varname></term>
+        <listitem>
+          <para>UDP encapsulation encompasses the techniques and protocols to
+          encapsulate and decapsulate networking packets of various protocols
+          inside UDPSupports below specific UDP encapsulation methods.
+          <literal>FooOverUDP</literal> <varname>FOU</varname> provides the simplest no
+          frills model of UDP encapsulation, it simply encapsulates packets directly in the UDP payload for
+          <literal>GenericUDPEncapsulation</literal> <varname>GUE</varname> is a generic and extensible
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Dot missing before "GUE".

We need some ulinks to documentation of GUE here.

File: man/systemd.netdev.xml [hunk 12]
```diff
@@ -747,8 +750,80 @@
...
+          encapsulation, it allows encapsulation of packets for any IP protocol and optional data as
+          part of the encapsulation. Default to <literal>FooOverUDP</literal>.
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Default*s*

File: src/network/netdev/netdev-gperf.gperf [hunk 13]
```diff
@@ -26,97 +26,104 @@ struct ConfigPerfItem;
...
+Bridge.ForwardDelaySec,            config_parse_sec,                     0,                             offsetof(Bridge, forward_delay)
+Bridge.Priority,                   config_parse_uint16,                  0,                             offsetof(Bridge, priority)
+Bridge.DefaultPVID,                config_parse_vlanid,                  0,                             offsetof(Bridge, default_pvid)
+Bridge.MulticastQuerier,           config_parse_tristate,                0,                             offsetof(Bridge, mcast_querier)
+Bridge.MulticastSnooping,          config_parse_tristate,                0,                             offsetof(Bridge, mcast_snooping)
+Bridge.VLANFiltering,              config_parse_tristate,                0,                             offsetof(Bridge, vlan_filtering)
+Bridge.STP,                        config_parse_tristate,                0,                             offsetof(Bridge, stp)
+VRF.TableId,                       config_parse_uint32,                  0,                             offsetof(Vrf, table_id)
+>>>>>>> networkd: add support for FooOverUDP and FOU to IPIP
```
Comment (keszybz) [inline src/network/netdev/netdev-gperf.gperf]: Whoa!

File: src/network/netdev/tunnel.c [hunk 14]
```diff
@@ -76,6 +81,34 @@ static int netdev_ipip_fill_message_create(NetDev *netdev, Link *link, sd_netlin
...
+
+                switch (t->fou_encap_type) {
+                case NETDEV_FOO_OVER_UDP_ENCAP_DIRECT:
+                        t->fou_encap_type = FOU_ENCAP_DIRECT;
+                        break;
+                case NETDEV_FOO_OVER_UDP_ENCAP_GUE:
+                        t->fou_encap_type = FOU_ENCAP_GUE;
+                        break;
+                default:
+                        t->fou_encap_type = FOU_ENCAP_DIRECT;
```
Comment (keszybz) [inline src/network/netdev/tunnel.c]: You're overwriting `fou_encap_type` with a different value. This looks wrong. Can you explain what is going on here?

File: src/network/netdev/tunnel.c [hunk 15]
```diff
@@ -600,6 +640,41 @@ int config_parse_encap_limit(const char* unit,
...
+        Tunnel *t = userdata;
+        uint16_t port;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (!STR_IN_SET(rvalue, "0", "auto")) {
```
Comment (keszybz) [inline src/network/netdev/tunnel.c]: I don't think there's any need to accept "0" in addition to "auto". Let's only permit "auto".

File: src/network/netdev/tunnel.c (hunk 14 above)
Comment (ssahani) [reply src/network/netdev/tunnel.c]: We need to have the same value as 
```
enum {
        FOU_ENCAP_UNSPEC,
        FOU_ENCAP_DIRECT,
        FOU_ENCAP_GUE,
};
```
defined in the fou.h . better way to do is below way . Will change.





```
    NETDEV_FOO_OVER_UDP_ENCAP_UNSPEC = FOU_ENCAP_UNSPEC,                                                                                                                                   
    NETDEV_FOO_OVER_UDP_ENCAP_DIRECT = FOU_ENCAP_DIRECT,                                                                                                                                   
    NETDEV_FOO_OVER_UDP_ENCAP_GUE = FOU_ENCAP_GUE, 
```

File: src/network/netdev/tunnel.c (hunk 15 above)
Comment (ssahani) [reply src/network/netdev/tunnel.c]: OK

File: src/libsystemd/sd-netlink/generic-netlink-message.c (hunk 2 above)
Comment (ssahani) [reply src/libsystemd/sd-netlink/generic-netlink-message.c]: OK

File: src/libsystemd/sd-netlink/generic-netlink-message.c (hunk 4 above)
Comment (ssahani) [reply src/libsystemd/sd-netlink/generic-netlink-message.c]: `CTRL_ATTR_MAX` is only for the control. We need to parse the attributes which is a variable size.
The caller should be knowing the size . For example ` ATTR types defined for L2TP' has `__L2TP_ATTR_MAX`  . We need to allocate the array size of that length.

File: configure.ac (hunk 1 above)
Comment (ssahani) [reply configure.ac]: Added.

File: src/libsystemd/sd-netlink/generic-netlink-message.c (hunk 3 above)
Comment (ssahani) [reply src/libsystemd/sd-netlink/generic-netlink-message.c]: Done

**Comment by ssahani** [comment]
Updated.

File: man/systemd.netdev.xml [hunk 16]
```diff
@@ -865,8 +868,74 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+          <term><varname>FooOverUDP=</varname></term>
+        <listitem>
+          <para>A boolean. Specifies whether <varname>FooOverUDP=</varname> tunnel is to be configured.
+           Default to false. For more detail information see
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Hmm, I'm pretty sure I pointed this out in preview review: there's a typo in "default".

File: man/systemd.netdev.xml [hunk 17]
```diff
@@ -865,8 +868,74 @@
...
+          <ulink url="https://lwn.net/Articles/614348">Foo over UDP</ulink></para>
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Some punctuation is missing at the end of the sentence.

File: man/systemd.netdev.xml [hunk 18]
```diff
@@ -865,8 +868,74 @@
...
+          packets arriving at <varname>FooOverUDPPort=</varname>. This field is mandatory and is not set
+          by default. Valid range is 1-255.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>EncapType=</varname></term>
+        <listitem>
+          <para>UDP encapsulation encompasses the techniques and protocols to
+          encapsulate and decapsulate networking packets of various protocols
+          inside UDPSupports below specific UDP encapsulation methods.
```
Comment (keszybz) [inline man/systemd.netdev.xml]: UDPSupports?

File: man/systemd.netdev.xml [hunk 19]
```diff
@@ -865,8 +868,74 @@
...
+          <literal>FooOverUDP</literal> <varname>FOU</varname> provides the simplest no
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Hmm, this paragraph is hard to parse. Can you make it a `varlist`, with each encapsulation type as a separate paragraph?

File: man/systemd.netdev.xml [hunk 20]
```diff
@@ -865,8 +868,74 @@
...
+          frills model of UDP encapsulation, it simply encapsulates packets directly in the UDP payload for
+          <literal>GenericUDPEncapsulation</literal>.<varname>GUE</varname> is a generic and extensible
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Space before GUE is missing.

File: man/systemd.netdev.xml [hunk 21]
```diff
@@ -865,8 +868,74 @@
...
+          part of the encapsulation. For more detail information see
+          <ulink url="https://lwn.net/Articles/615044">Generic UDP Encapsulation</ulink>.
+          Defaults to <literal>FooOverUDP</literal>.
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>Port=</varname></term>
+        <listitem>
+          <para>The interface name used when creating the netdev.
```
Comment (keszybz) [inline man/systemd.netdev.xml]: Is that the interface name or port number?

File: meson.build [hunk 22]
```diff
@@ -410,7 +410,10 @@ foreach decl : [['IFLA_INET6_ADDR_GEN_MODE',         'linux/if_link.h'],
                 ['NDA_IFINDEX',                      'linux/neighbour.h'],
                 ['IFA_FLAGS',                        'linux/if_addr.h'],
                 ['LO_FLAGS_PARTSCAN',                'linux/loop.h'],
-               ]
+                ['FOU_ATTR_REMCSUM_NOPARTIAL',       'linux/fou.h'],
+                ['FOU_CMD_GET',                      'linux/fou.h'],
+                ['FOU_ENCAP_GUE',                    'linux/fou.h'],
+              ]
```
Comment (keszybz) [inline meson.build]: This brace was aligned to the same column as the opening brace. Please leave it be.

File: src/libsystemd/sd-netlink/generic-netlink-message.c (hunk 3 above)
Comment (keszybz) [reply src/libsystemd/sd-netlink/generic-netlink-message.c]: Once again: `sd_netlin_message_append_data` does some of the checks. The only one that it *doesn't* do is the one on `data`. So you can simplify this function to make it a simple wrapper like I suggested above.

File: src/libsystemd/sd-netlink/generic-netlink-message.c [hunk 23]
```diff
@@ -0,0 +1,247 @@
...
+
+        length = strlen(data);
+
+        return sd_netlink_message_append_data(m, type, data, length + 1);
+}
+
+int sd_generic_netlink_message_append_u8(sd_netlink_message *m, unsigned short type, uint8_t data) {
+
+        assert_return(m, -EINVAL);
+        assert_return(!m->sealed, -EPERM);
```
Comment (keszybz) [inline src/libsystemd/sd-netlink/generic-netlink-message.c]: The same applies here and in the other similar functions below.

File: src/network/netdev/fou-tunnel.c [hunk 24]
```diff
@@ -0,0 +1,180 @@
...
+
+        assert(t);
+
+        r = sd_rtnl_message_new_generic_netlink(id, FOU_CMD_ADD, FOU_GENL_VERSION, &m);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Failed to create generic netlink message: %m");
+
+        r = sd_generic_netlink_message_append_u16(m, FOU_ATTR_PORT, htobe16(t->port));
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not append FOU_ATTR_PORT, attribute: %m");
```
Comment (keszybz) [inline src/network/netdev/fou-tunnel.c]: Spurious comma.

File: src/network/netdev/fou-tunnel.c [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+
+        assert(netdev);
+
+        t = FOU(netdev);
+
+        assert(t);
+
+        r = sd_generic_netlink_resolve_family(netdev->manager->generic_rtnl, FOU_GENL_NAME, &id);
+        if (r < 0)
+                return log_netdev_error_errno(netdev, r, "Could not reolve '%s' family name to id: %m", FOU_GENL_NAME);
```
Comment (keszybz) [inline src/network/netdev/fou-tunnel.c]: Typo.

File: src/network/netdev/fou-tunnel.c [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+
+        r = netdev_fill_fou_tunnel_message(netdev, id, &m);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_call(netdev->manager->generic_rtnl, m, 0, NULL);
+        if (r < 0 && r != -EADDRINUSE)
+                return log_netdev_error_errno(netdev, r, "Failed to add FooOverUDP tunnel: %m");
```
Comment (keszybz) [inline src/network/netdev/fou-tunnel.c]: Well, if we fail because the port is busy, that's not good. Normally, we fail to create something because it already exists, and that is OK because creation of network devices requires privileges and they cannot be created by mistake. But ports can be opened by unprivileged users and also programs running as root, so port conflicts can happen. This failure should not be silent.

File: src/network/netdev/netdev-gperf.gperf [hunk 27]
```diff
@@ -57,6 +58,12 @@ Tunnel.IPv6FlowLabel,        config_parse_ipv6_flowlabel,          0,
 Tunnel.CopyDSCP,             config_parse_bool,                    0,                             offsetof(Tunnel, copy_dscp)
 Tunnel.EncapsulationLimit,   config_parse_encap_limit,             0,                             offsetof(Tunnel, encap_limit)
 Tunnel.Independent,          config_parse_bool,                    0,                             offsetof(Tunnel, independent)
+Tunnel.FOU,                  config_parse_bool,                    0,                             offsetof(Tunnel, fou_tunnel)
+Tunnel.FOUDestinationPort,   config_parse_tunnel_fou_port,         0,                             0
+Tunnel.FOUSourcePort,        config_parse_tunnel_fou_port,         0,                             0
+FouOverUDP.Protocol,         config_parse_uint16,                  0,                             offsetof(FouTunnel, fou_protocol)
```
Comment (keszybz) [inline src/network/netdev/netdev-gperf.gperf]: FOU == "Foo Over UDP". What does "FouOverUDP" stand for?

File: man/systemd.netdev.xml (hunk 16 above)
Comment (ssahani) [reply man/systemd.netdev.xml]: Defaults ?

File: src/network/netdev/fou-tunnel.c (hunk 26 above)
Comment (ssahani) [reply src/network/netdev/fou-tunnel.c]: This is interesting. 
1. First time we create this port via networkd.
2. now we restart. we get into this situation because this port does not get removed if shut or restart networkd.

File: src/network/netdev/netdev-gperf.gperf (hunk 27 above)
Comment (ssahani) [reply src/network/netdev/netdev-gperf.gperf]: Ouch

File: man/systemd.netdev.xml (hunk 21 above)
Comment (ssahani) [reply man/systemd.netdev.xml]: port number will fix this.

File: man/systemd.netdev.xml (hunk 19 above)
Comment (ssahani) [reply man/systemd.netdev.xml]: `EncapType=` could be `FOU`/`GUE`. May be I will try with simpler.

**Comment by ssahani** [comment]
Thanks for investing so much of time on this. I have updated. Guess have not missed anything.

**Comment by ssahani** [comment]
@keszybz Ping !

**Comment by ssahani** [comment]
@keszybz anything waiting on me . Couple of features wating on for generic netlink
https://github.com/systemd/systemd/issues/5754 and L2TP which I have the patch ready.

**Comment by yuwata** [comment]
@ssahani At least, now this needs to rebase.

**Comment by ssahani** [comment]
Updated

File: man/systemd.netdev.xml [hunk 28]
```diff
@@ -897,8 +900,75 @@
...
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>EncapType=</varname></term>
+        <listitem>
+          <para>UDP encapsulation encompasses the techniques and protocols to
+          encapsulate and decapsulate networking packets of various protocols
+          inside UDP. Supports below specific UDP encapsulation methods.
+
+          <varname>FOU</varname> provides the simplest no frills model of UDP encapsulation, it simply encapsulates
```
Comment (yuwata) [inline man/systemd.netdev.xml]: Please use literal tag.

File: man/systemd.netdev.xml [hunk 29]
```diff
@@ -897,8 +900,75 @@
...
+          packets directly in the UDP payload.
+          <varname>GUE</varname> is a generic and extensible encapsulation, it allows encapsulation of packets for any IP
```
Comment (yuwata) [inline man/systemd.netdev.xml]: Please use literal tag.

File: man/systemd.netdev.xml [hunk 30]
```diff
@@ -897,8 +900,75 @@
...
+          protocol and optional data as part of the encapsulation.
+          For more detail information see <ulink url="https://lwn.net/Articles/615044">Generic UDP Encapsulation</ulink>.
+          Defaults to <varname>FOU</varname>.
```
Comment (yuwata) [inline man/systemd.netdev.xml]: please use literal tag.

File: man/systemd.netdev.xml [hunk 31]
```diff
@@ -1295,8 +1365,34 @@ Remote=192.169.224.239
 TTL=64</programlisting>
     </example>
     <example>
-      <title>/etc/systemd/network/25-tap.netdev</title>
-      <programlisting>[NetDev]
+<title>/etc/systemd/network/1-fou-tunnel.netdev</title>
```
Comment (yuwata) [inline man/systemd.netdev.xml]: please fix indentation.

File: man/systemd.netdev.xml [hunk 32]
```diff
@@ -897,8 +900,75 @@
...
+          </para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>Port=</varname></term>
+        <listitem>
+          <para> Specifies the port number, where the IP encapsulation packets will arrive. Please take note that the packets
+          will arrive with the encapsulation will be removed. Then they will be manually fed back into the network stack, and sent ahead
+          for delivery to the real destination.
+          This option is compulsory.</para>
```
Comment (yuwata) [inline man/systemd.netdev.xml]: I prefer 'mandatory'.

File: src/network/netdev/netdev-gperf.gperf [hunk 33]
```diff
@@ -64,6 +65,12 @@ Tunnel.CopyDSCP,             config_parse_bool,                    0,
 Tunnel.EncapsulationLimit,   config_parse_encap_limit,             0,                             offsetof(Tunnel, encap_limit)
 Tunnel.Independent,          config_parse_bool,                    0,                             offsetof(Tunnel, independent)
 Tunnel.AllowLocalRemote,     config_parse_tristate,                0,                             offsetof(Tunnel, allow_localremote)
+Tunnel.FooOverUDP,           config_parse_bool,                    0,                             offsetof(Tunnel, fou_tunnel)
+Tunnel.FOUDestinationPort,   config_parse_tunnel_fou_port,         0,                             0
+Tunnel.FOUSourcePort,        config_parse_tunnel_fou_port,         0,                             0
+FooOverUDP.Protocol,         config_parse_uint16,                  0,                             offsetof(FouTunnel, fou_protocol)
+FooOverUDP.EncapType,        config_parse_fou_encap_type,          0,                             offsetof(FouTunnel, fou_encap_type)
+FooOverUDP.Port,             config_parse_fou_port,                0,                             0
```
Comment (yuwata) [inline src/network/netdev/netdev-gperf.gperf]: Why do not you use `config_parse_ip_port()`?

File: src/network/netdev/tunnel.c [hunk 34]
```diff
@@ -613,13 +640,49 @@ int config_parse_encap_limit(const char* unit,
...
+        Tunnel *t = userdata;
+        uint16_t port = 0;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        if (!streq(rvalue, "auto")) {
```
Comment (yuwata) [inline src/network/netdev/tunnel.c]: Taking 'auto' is necessary? When 'auto' is specified, then the port variable is set to 0. But this is the default value. So, we do not need to set 'auto' to SourcePort=...
I think config_parse_ip_port() is sufficient for these settings.

File: man/systemd.netdev.xml [hunk 35]
```diff
@@ -897,8 +900,75 @@
...
+        <listitem>
+          <para>The <varname>FOUDestinationPort=</varname> specifies the UDP destination port for encapsulation.
+          This field is mandatory and is not set by default.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>FOUSourcePort=</varname></term>
+        <listitem>
+          <para>The <varname>FOUSourcePort=</varname> specifies the
+          UDP source port for encapsulation. Defaults to <varname>auto</varname>,
```
Comment (yuwata) [inline man/systemd.netdev.xml]: Should this support the special value 'auto'?
At least, please use literal tag.

File: man/systemd.netdev.xml [hunk 36]
```diff
@@ -897,8 +900,75 @@
...
+    netdevs of kind <literal>fou</literal> and accepts the
+    following keys:</para>
+
+    <variablelist class='network-directives'>
+      <varlistentry>
+        <term><varname>Protocol=</varname></term>
+        <listitem>
+          <para>The <varname>Protocol=</varname> specifies the protocol number of the
+          packets arriving at <varname>FooOverUDPPort=</varname>. This field is mandatory and is not set
+          by default. Valid range is 1-255.</para>
```
Comment (yuwata) [inline man/systemd.netdev.xml]: When encap type is FOU, then protocol is mandatory. However, protocol cannot specified when GUE is used.

File: man/systemd.netdev.xml (hunk 28 above)
Comment (yuwata) [reply man/systemd.netdev.xml]: FOU and GUE are not valid option.

File: src/network/netdev/fou-tunnel.c [hunk 37]
```diff
@@ -0,0 +1,180 @@
...
+                log_syntax(unit, LOG_ERR, filename, line, r, "Failed to parse tunnel's '%s' port '%s'.", lvalue, rvalue);
+                return 0;
+        }
+
+        t->port = port;
+
+        return 0;
+}
+
+static int netdev_fou_tunnel_verify(NetDev *netdev, const char *filename) {
```
Comment (yuwata) [inline src/network/netdev/fou-tunnel.c]: needs to check that port is set.

File: src/network/netdev/tunnel.h [hunk 38]
```diff
@@ -59,10 +59,15 @@ typedef struct Tunnel {
         union in_addr_union remote;
 
         Ip6TnlMode ip6tnl_mode;
+        FooOverUDPEncapType fou_encap_type;
```
Comment (yuwata) [inline src/network/netdev/tunnel.h]: is this used?

File: src/network/netdev/netdev-gperf.gperf (hunk 33 above)
Comment (ssahani) [reply src/network/netdev/netdev-gperf.gperf]: yes

File: src/network/netdev/tunnel.c (hunk 34 above)
Comment (ssahani) [reply src/network/netdev/tunnel.c]: yes

**Comment by yuwata** [comment]
(Revised version is #9170)
